### ISSUE TITLE
mysofa_open loading freezes: infinite while loop in kdtree clear_results() method

### ISSUE THREAD

**Issue by PyrApple** [hoene/libmysofa#57]
# Description
Successive loading of various sofa HRIR files in an audio plugin. Loading goes smooth for the first 10 to 20+ loadings (user interaction based), then will freeze while trying to load a file. Not a specific file, usually just going back and forth on 2-3 HRIRs (loading successfully otherwise) will produce the freeze at some point.

# Detail
Freeze occurs in ```clear_results(...)``` method of kdtree.c, during the while loop:
```
while (node) {
    tmp = node;
    node = node->next;
    free_resnode(tmp);
}
```
Every time I reached the frozen state, ```node->next``` and ```free_nodes``` were always equal (and not ```NULL```), hence in ```free_resnode(...)``` the 
```
node->next = free_nodes;
free_nodes = node;
```
double lines eventually led to a ```node`` which ```next``` attribute was itself (and so forth). The ```while(node)``` condition of the ```clear_results(...)``` method thus was never fulfilled, hence the infinite loop.

If ever it rings a bell somewhere, I'll take it. Sorry I was not able to produce a minimalist test scenario.

**Comment by hoene** [comment]
May you double check your program with valgrind?

**Comment by hoene** [comment]
@jtsiomb John, do you have any idea on what the problem could be?

**Comment by jtsiomb** [comment]
It does sound like a memory bug, or more likely a double-free situation. If you try to free the same result set twice, the same nodes will end up in the free list twice, and you can easily end up with cycles in the list.

You can try to side-step the issue by disabling the fast node allocator with ./configure --disable-fastalloc. If you're not building kdtree separately, that configure option just amounts to removing the USE_LIST_NODE_ALLOCATOR definition. But if it was a double-free issue then you'll end up calling free() twice on the same node, which is equally problematic, and if it's another kind of memory corruption, it will just move on to affect something else, possibly intermittently. So it's best if you figure it out now. Valgrind is a good suggestion, if a code inspection doesn't reveal anything obvious.

If all else fails I might be able to look into it, but only if you manage to reproduce the problem in a minimal program that only uses libkdtree.

**Comment by hoene** [comment]
Hi, sorry. Currently, your bug is not yet located precisely.  Something strange seems to happen. 
Sadly enough, in C this all but to frequent because of memory overwrite.  Valgrind helps in those cases but if it does not work with VST, that is bad needs. I do not know any alternative (you might try one of these tools: https://alternativeto.net/software/valgrind/)

**Comment by PyrApple** [comment]
Managed to reproduce the [bug as a standalone](https://www.dropbox.com/s/zyzqkcz1ujndigj/sofaCyclicFreeDebug.zip?dl=0): 2 threads reproducing the setup described in previous comment, program freezes when both run simultaneously.
Will give it a try with Valgrind asa I've got time on my hands.

**Comment by PyrApple** [comment]
Good news, the standalone above does not raise any bug on Windows (10) and Ubuntu (16.04). Bad news, [valgrind segfaults on osx pthread](https://stackoverflow.com/questions/44990380/why-valgrind-segfaults-when-i-launch-a-new-thread), preventing any further tests on mysofa multi-thread access atm.

**Comment by PyrApple** [comment]
Hi Christian, John. The bug was introduced by the (latest as for today) commit https://github.com/hoene/libmysofa/commit/5e1fc5b62d6d43cbac123e18dfee61408705c4a1. Continuing investigations :).

**Comment by hoene** [comment]
Okay, so we have two thread accesssing libmysofa at the same time?

**Comment by PyrApple** [comment]
Yes. In the debug project attached above, two threads instantiate each their own ```MYSOFA_EASY struct```. One loops over ```mysofa_getfilter_float``` method, the other over ```mysofa_open```.
